src/Microsoft.DotNet.Interactive/Server/KernelCommandEnvelope.cs

```diff
@@ -17,11 +17,40 @@ public abstract class KernelCommandEnvelope : IKernelCommandEnvelope
         private static readonly ConcurrentDictionary<Type, Func<IKernelCommand, IKernelCommandEnvelope>> _envelopeFactories =
             new ConcurrentDictionary<Type, Func<IKernelCommand, IKernelCommandEnvelope>>();
 
-        private static readonly Dictionary<string, Type> _envelopeTypesByCommandTypeName;
+        private static Dictionary<string, Type> _envelopeTypesByCommandTypeName;
 
-        private static readonly Dictionary<string, Type> _commandTypesByCommandTypeName;
+        private static Dictionary<string, Type> _commandTypesByCommandTypeName;
 
+      
         static KernelCommandEnvelope()
+        {
+            ResetToDefaults();
+        }
+
+        internal static Type CommandTypeByName(string name) => _commandTypesByCommandTypeName[name];
+
+        private readonly IKernelCommand _command;
+
+        protected KernelCommandEnvelope(IKernelCommand command)
+        {
+            _command = command ?? throw new ArgumentNullException(nameof(command));
+        }
+
+        public abstract string CommandType { get; }
+
+        public string Token => _command.GetToken();
+
+        IKernelCommand IKernelCommandEnvelope.Command => _command;
+
+        public static void RegisterCommandType<T>(string commandTypeName) where T : class, IKernelCommand
+        {
+            var commandEnvelopeType = typeof(KernelCommandEnvelope<T>);
+            var commandType = typeof(T);
+
+            _envelopeTypesByCommandTypeName[commandTypeName] = commandEnvelopeType;
```

Comment:
I wonder if we want to use `Add` so that it throws if the command type is already added.

Reply:
I think is a good point